src/services/refactors/inlineVariable.ts

```diff
@@ -0,0 +1,196 @@
+        const info = getInliningInfo(file, startPosition, triggerReason === "invoked", program);
+        if (!info) {
+            return emptyArray;
+        }
+
+        if (refactor.isRefactorErrorInfo(info) && preferences.provideRefactorNotApplicableReason) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [{
+                    ...inlineVariableAction,
+                    notApplicableReason: info.error
+                }]
+            }];
+        }
+
+        if (!refactor.isRefactorErrorInfo(info)) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [inlineVariableAction]
+            }];
+        }
+
+        return emptyArray;
+    },
+
+    getEditsForAction(context, actionName) {
+        Debug.assert(actionName === refactorName, "Unexpected refactor invoked");
+
+        const { file, program, startPosition } = context;
+
+        const info = getInliningInfo(file, startPosition, /*tryWithReferenceToken*/ true, program);
+        if (!info || refactor.isRefactorErrorInfo(info)) {
+            return undefined;
+        }
+
+        const { references, declaration, replacement } = info;
+        const edits = textChanges.ChangeTracker.with(context, tracker => {
+            for (const node of references) {
+                tracker.replaceNode(file, node, getReplacementExpression(node, replacement));
+            }
+            tracker.delete(file, declaration);
+        });
+
+        return { edits };
+    }
+});
+
+function getInliningInfo(file: SourceFile, startPosition: number, tryWithReferenceToken: boolean, program: Program): InliningInfo | RefactorErrorInfo | undefined {
+    const checker = program.getTypeChecker();
+    const token = getTokenAtPosition(file, startPosition);
+    const parent = token.parent;
+
+    // If the node is a variable declaration, make sure it's not in a catch clause or for-loop
+    // and that it has a value.
+    if (isInitializedVariable(parent) && isVariableDeclarationInVariableStatement(parent) && isIdentifier(parent.name)) {
+        // Don't inline the variable if it has multiple declarations.
+        if (parent.symbol.declarations?.length !== 1) {
+            return { error: getLocaleSpecificMessage(Diagnostics.Variables_that_share_a_name_with_a_type_or_namespace_in_the_same_scope_cannot_be_inlined) };
+        }
+
+        // Do not inline if the variable is exported.
+        if (findAncestor(parent, node => canHaveModifiers(node) && some(node.modifiers, isExportModifier))) {
+            return undefined;
+        }
+
+        // Find all references to the variable in the current file.
+        const references = getReferenceNodes(parent, checker, file);
+        return references && { references, declaration: parent, replacement: getSynthesizedDeepClone(parent.initializer) };

```

Comment:
Okay, so this is the subtlety that I was talking about in (https://github.com/microsoft/TypeScript/pull/54281#discussion_r1212415990).

It is not good enough to just get the synthesized deep clone once. You need to do it for each reference. Otherwise, that clone gets shared among each replacement, and that's where things will often go wrong.

Currently, the easiest way to do this is just to create the clone before passing it to `getReplacementExpression`.

That said, I'm still not able to come up with a repro test case.

Reply:
Ohhh got it. I'll make sure that each reference gets its own replacement clone.